### PR TITLE
(dummy) Scheduler implementation

### PR DESCRIPTION
**Pull request by capossele**
This PR adds the (dummy) scheduler implementation. However, the scheduler must still be wired into the Tangle flow.

### CODE REVIEW

File: packages/tangle/scheduler.go [hunk 1]
```diff
@@ -0,0 +1,299 @@
...
+					s.timeQueue.Add(message)
+					break
+				}
+				s.trySchedule(message)
+
+			// try to schedule messsage that have been waiting and are now current.
+			case message := <-s.timeQueue.C:
+				s.trySchedule(message)
+
+			// schedule messages that were waiting for of their parents to be booked.
```
Comment (jkrvivian) [inline packages/tangle/scheduler.go:101]:
```suggestion
			// schedule messages that were waiting for their parents to be booked.
```

File: packages/tangle/scheduler.go [hunk 2]
```diff
@@ -0,0 +1,298 @@
...
+
+// Scheduler implements the scheduler.
+type Scheduler struct {
+	Events *SchedulerEvents
+
+	tangle *Tangle
+
+	inbox            chan *MessageToSchedule
+	timeQueue        *TimeMessageQueue
+	parentsMap       SchedulerParentPriorityMap
```
Comment (lzpap) [inline packages/tangle/scheduler.go:37]: should this be protected with a mutex?
we check on incoming message of parents have been booked, but the multithreaded outbox worker could modifiy the map in the meantime, no?

File: packages/tangle/scheduler.go [hunk 3]
```diff
@@ -0,0 +1,298 @@
...
+	}
+
+	return parents
+}
+
+// region TimeMessageQueue /////////////////////////////////////////////////////////////////////////////////////////////
+
+// TimeMessageQueue is a time-based ordered queue.
+type TimeMessageQueue struct {
+	list    timeIssuanceSortedList
```
Comment (lzpap) [inline packages/tangle/scheduler.go:175]: what about this list? should this be protected as well?

File: packages/tangle/tangle.go [hunk 4]
```diff
@@ -44,10 +45,12 @@ func New(options ...Option) (tangle *Tangle) {
 	tangle.Parser = NewParser()
 	tangle.Storage = NewStorage(tangle)
 	tangle.Solidifier = NewSolidifier(tangle)
+	tangle.Scheduler = NewScheduler(tangle)
+	tangle.LedgerState = NewLedgerState(tangle)
+	tangle.Booker = NewBooker(tangle)
```
[the remark this answers is not in the document]
Comment (capossele) [reply packages/tangle/tangle.go:50]: Yeah, I think the LesgerState could also be not initialised, while instead the Booker initialisation is necessary as the scheduler attach to one of its events. However, is the setup phase that will do the proper wiring

File: packages/tangle/scheduler.go (hunk 2 above)
Comment (capossele) [reply packages/tangle/scheduler.go:37]: No that map is managed entirely by the scheduler

File: packages/tangle/scheduler.go (hunk 3 above)
Comment (capossele) [reply packages/tangle/scheduler.go:175]: Also managed only by the scheduler

Comment (capossele) [reply packages/tangle/scheduler.go:175]: I mean, it is internally used by the TimeMessageQueue, which is managed by the scheduler and its access is restricted via a select

**Review by lzpap — APPROVED** [review]

**Pull request merged**